lib/Tapable.ts

```diff
@@ -2,40 +2,120 @@
 	MIT License http://www.opensource.org/licenses/mit-license.php
 	Author Tobias Koppers @sokra
 */
-function Tapable() {
+
+// interface Tapable {
+//    compilation(compilation, data): void;
+// }
+
+// this.plugin("compilation", function (compilation, data) { ... });
+
+interface Plugin {
+	apply(): void; 
+}
+
+export abstract class Tapable {
+	private _plugins: any = {};
+	static mixin(property: string): void {
+		copyProperties(Tapable.prototype, property);	 
+	}
+
+	public apply(...plugins: Plugin[]) {
+		for(let i = 0; i < plugins.length; i++) {
+			plugins[i].apply(this);
+		}
+	}
+
+	public plugin(names: string|string[], fn: Function ) {
+		if(Array.isArray(names)) {
+			names.forEach(function(name: string){
+				this.plugin(name, fn);
+			}, this);
+			return;
+		}
+		if(!this._plugins[name]) this._plugins[name] = [fn];
+		else this._plugins[name].push(fn);
+	}
+
+	protected applyPlugins(name: string, ...args: any[]) {
+		if(!this._plugins[name]) return;
+		
+		let plugins = this._plugins[name];
+		for(let i = 0; i < plugins.length; i++)
+			plugins[i].apply(this, args);
+	}
+
+	protected applyPluginsWaterfall(name: string, init: any, ...args: any[]) {
+		if(!this._plugins[name]) return init;
+
+		let plugins = this._plugins[name];
+		let current = init;
+		for(let i = 0; i < plugins.length; i++)
+			current = plugins[i].apply(this, [current].concat(args));
+		return current;
+	}
+
+	protected applyPluginsWaterfall0(name: string, init: any) {
```

Comment:
So changing the actual name of the function would require a huge changes to webpack/core, but once I've gotten this fully migrated, I can add a comment to explain how some are ussing `call` and some use `apply` thus the strange name.